### PR TITLE
Feature 2163 new review period

### PR DESCRIPTION
**Pull request by mvolkmann**

### CODE REVIEW

File: web-ui/.prettierrc [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{
```
Comment (mvolkmann) [inline web-ui/.prettierrc:1]: Adding configuration for Prettier.

File: web-ui/src/components/reviews/TeamReviews.jsx [hunk 2]
```diff
@@ -1,34 +1,44 @@
-import React, {useEffect, useContext, useCallback, useState, useRef} from "react";
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/TeamReviews.jsx:1]: Nearly all the changes in this file are from Prettier formatting.

File: web-ui/src/components/reviews/TeamReviews.jsx [hunk 3]
```diff
@@ -37,47 +47,55 @@ import {
...
-    },
+    margin: '0.5em 0 0 1em',
+    ['@media (max-width:820px)']: {
+      // eslint-disable-line no-useless-computed-key
+      padding: '0'
+    }
   },
   [`& .${classes.headerContainer}`]: {
-    display: "flex",
-    'flex-direction': "row",
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/TeamReviews.jsx:68]: I changed all CSS properties that do not use camelCase names to use them in order to remove warnings.

File: web-ui/src/pages/ReviewsPage.jsx [hunk 4]
```diff
@@ -1,43 +1,46 @@
-import React, {useCallback, useEffect, useState} from "react";
-import { styled } from '@mui/material/styles';
```
Comment (mvolkmann) [inline web-ui/src/pages/ReviewsPage.jsx:2]: Nearly all the changes in this file are from Prettier formatting. I'll add comments below that describe the real changes.

File: web-ui/src/pages/ReviewsPage.jsx [hunk 5]
```diff
@@ -1,43 +1,46 @@
...
-  stepContainer: `${PREFIX}-stepContainer`,
+  stepContainer: `${PREFIX}-stepContainer`
 };
 
 const Root = styled('div')(({theme}) => ({
   [`&.${classes.root}`]: {
-    backgroundColor: "transparent",
-    margin: "4rem 2rem 2rem 2rem",
-    height: "100%",
-    'max-width': "100%",
```
Comment (mvolkmann) [inline web-ui/src/pages/ReviewsPage.jsx:21]: I changed all CSS properties that do not use camelCase names to use them in order to remove warnings.

File: web-ui/src/pages/SelfReviewsPage.jsx [hunk 6]
```diff
@@ -1,57 +1,61 @@
-import React, {useContext, useCallback, useEffect, useState} from "react";
```
Comment (mvolkmann) [inline web-ui/src/pages/SelfReviewsPage.jsx:1]: Nearly all the changes in this file are from Prettier formatting. I'll add comments below that describe the real changes.

File: web-ui/src/pages/SelfReviewsPage.jsx [hunk 7]
```diff
@@ -1,57 +1,61 @@
...
-  stepContainer: `${PREFIX}-stepContainer`,
+  stepContainer: `${PREFIX}-stepContainer`
 };
 
 const Root = styled('div')(({theme}) => ({
   [`&.${classes.root}`]: {
-    backgroundColor: "transparent",
-    padding: "4rem 2rem 0 2rem",
-    height: "100%",
-    'max-width': "100%",
```
Comment (mvolkmann) [inline web-ui/src/pages/SelfReviewsPage.jsx:24]: I changed all CSS properties that do not use camelCase names to use them in order to remove warnings.

File: web-ui/src/components/reviews/periods/ReviewPeriods.jsx [hunk 8]
```diff
@@ -1,162 +1,204 @@
-import React, {useCallback, useContext, useEffect, useState} from "react";
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/periods/ReviewPeriods.jsx:1]: Many changes in this file are from Prettier formatting.
I also sorted the imports a bit.
For example, all the imports from "@mui/icons-material" are together
and all the imports from "@mui/material" are together.
I'll add comments below that describe all the changes not related to Prettier formatting.

File: web-ui/src/components/reviews/periods/ReviewPeriods.jsx [hunk 9]
```diff
@@ -1,162 +1,204 @@
...
+import DialogContentText from '@mui/material/DialogContentText';
+import DialogTitle from '@mui/material/DialogTitle';
+import FormControl from '@mui/material/FormControl';
+import IconButton from '@mui/material/IconButton';
+import InputLabel from '@mui/material/InputLabel';
 import List from '@mui/material/List';
 import ListItem from '@mui/material/ListItem';
 import ListItemText from '@mui/material/ListItemText';
 import ListItemAvatar from '@mui/material/ListItemAvatar';
+import MenuItem from '@mui/material/MenuItem';
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/periods/ReviewPeriods.jsx:24]: This MUI component wasn't previously used here, but is now.

File: web-ui/src/components/reviews/periods/ReviewPeriods.jsx [hunk 10]
```diff
@@ -1,162 +1,204 @@
...
 import Modal from '@mui/material/Modal';
+import Select from '@mui/material/Select';
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/periods/ReviewPeriods.jsx:26]: This MUI component wasn't previously used here, but is now.

File: web-ui/src/components/reviews/periods/ReviewPeriods.jsx [hunk 11]
```diff
@@ -1,162 +1,204 @@
...
+  minWidth: '400px',
+  maxWidth: '600px',
+  backgroundColor: 'background.paper',
+  top: '50%',
+  left: '50%',
+  padding: '1rem',
+  transform: 'translate(-50%, -50%)',
+  border: '2px solid #fff',
+
+  display: 'flex',
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/periods/ReviewPeriods.jsx:80]: I added these three CSS properties to get better vertical spacing between the components.

File: web-ui/src/components/reviews/periods/ReviewPeriods.jsx [hunk 12]
```diff
@@ -1,162 +1,204 @@
...
-  const csrf = selectCsrfToken(state);
-  const periods = selectReviewPeriods(state);
-  const [selfReviews, setSelfReviews] = useState(null);
-  const [periodToAdd, setPeriodToAdd] = useState({ name: "", open: true });
-  const isAdmin = userProfile?.role?.includes("ADMIN");
-  const [open, setOpen] = useState(false)
+const ReviewPeriods = ({onPeriodSelected, mode}) => {
+  const {state, dispatch} = useContext(AppContext);
+
+  const [canSave, setCanSave] = useState(false);
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/periods/ReviewPeriods.jsx:112]: This is a new bit of state that is used to conditionally disable the Save button.

File: web-ui/src/components/reviews/periods/ReviewPeriods.jsx [hunk 13]
```diff
@@ -1,162 +1,204 @@
...
   const [confirmOpen, setConfirmOpen] = useState(false);
   const [loading, setLoading] = useState(false);
+  const [open, setOpen] = useState(false);
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/periods/ReviewPeriods.jsx:115]: All the `useState` calls are now grouped together and alphabetized.

File: web-ui/src/components/reviews/periods/ReviewPeriods.jsx [hunk 14]
```diff
@@ -168,15 +210,33 @@ const ReviewPeriods = ({ onPeriodSelected, mode }) => {
...
-    dispatch({ type: UPDATE_REVIEW_PERIODS, payload: periods.filter((period) => period?.id !== toDelete) });
+    await removeReviewPeriod(toDelete, csrf);
+    dispatch({
+      type: UPDATE_REVIEW_PERIODS,
+      payload: periods.filter(period => period?.id !== toDelete)
+    });
     handleConfirmClose();
   }, [csrf, periods, dispatch, toDelete, handleConfirmClose]);
 
+  const loadFeedbackTemplates = useCallback(async () => {
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/periods/ReviewPeriods.jsx:221]: This function is new. It gets all the available feedback templates so they can be displayed in drop-downs.

File: web-ui/src/components/reviews/periods/ReviewPeriods.jsx [hunk 15]
```diff
@@ -168,15 +210,33 @@ const ReviewPeriods = ({ onPeriodSelected, mode }) => {
...
+    const res = await getAllFeedbackTemplates(csrf);
+    const templates = res.payload.data;
+    templates.sort((t1, t2) => t1.title.localeCompare(t2.title));
+    setTemplates(templates);
+  }, [csrf, dispatch]);
+
+  useEffect(() => {
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/periods/ReviewPeriods.jsx:228]: This `useEffect` is new. It determines whether the Save button should be disabled based on the values entered and selected by the user.

File: web-ui/src/components/reviews/periods/ReviewPeriods.jsx [hunk 16]
```diff
@@ -168,15 +210,33 @@ const ReviewPeriods = ({ onPeriodSelected, mode }) => {
...
+    const valid = Boolean(
+      periodToAdd.name &&
+        periodToAdd.reviewTemplateId &&
+        periodToAdd.selfReviewTemplateId
+    );
+    setCanSave(valid);
+  }, [periodToAdd]);
+
   useEffect(() => {
+    loadFeedbackTemplates();
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/periods/ReviewPeriods.jsx:238]: Calling the new function to load the feedback templates.

File: web-ui/src/components/reviews/periods/ReviewPeriods.jsx [hunk 17]
```diff
@@ -199,90 +259,165 @@ const ReviewPeriods = ({ onPeriodSelected, mode }) => {
...
+      setConfirmOpen(true);
+    },
+    [setToDelete, setConfirmOpen]
+  );
 
-  const confirmDelete = useCallback((id) => {
-    setToDelete(id);
-    setConfirmOpen(true);
-  }, [setToDelete, setConfirmOpen]);
+  const handleReviewTemplateChange = event => {
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/periods/ReviewPeriods.jsx:313]: This is a new function that is called when the user selected a Review Template.

File: web-ui/src/components/reviews/periods/ReviewPeriods.jsx [hunk 18]
```diff
@@ -199,90 +259,165 @@ const ReviewPeriods = ({ onPeriodSelected, mode }) => {
...
+    const templateId = event.target.value;
+    setPeriodToAdd({
+      ...periodToAdd,
+      reviewTemplateId: templateId
+    });
+  };
+
+  const handleSelfReviewTemplateChange = event => {
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/periods/ReviewPeriods.jsx:321]: This is a new function that is called when the user selected a Self-Review Template.

File: web-ui/src/components/reviews/periods/ReviewPeriods.jsx [hunk 19]
```diff
@@ -292,38 +427,72 @@ const ReviewPeriods = ({ onPeriodSelected, mode }) => {
             label="Name"
             placeholder="Period Name"
             required
-            value={periodToAdd.name ? periodToAdd.name : ""}
-            onChange={(e) =>
-              setPeriodToAdd({ ...periodToAdd, name: e.target.value })
-            }
-          />
-          <TextField
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/periods/ReviewPeriods.jsx:300]: This TextField was replaced by a Select component that is wrapped in a FormControl.

File: web-ui/src/components/reviews/periods/ReviewPeriods.jsx [hunk 20]
```diff
@@ -292,38 +427,72 @@ const ReviewPeriods = ({ onPeriodSelected, mode }) => {
...
-            id="reviewPeriod-template-input"
-            label="Template ID"
-            placeholder="Template UUID"
-            required
-            value={periodToAdd.reviewTemplateId ? periodToAdd.reviewTemplateId : ""}
-            onChange={(e) =>
-              setPeriodToAdd({ ...periodToAdd, reviewTemplateId: e.target.value })
-            }
-          />
-          <TextField
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/periods/ReviewPeriods.jsx:311]: This TextField was replaced by a Select component that is wrapped in a FormControl.

File: web-ui/src/components/reviews/periods/ReviewPeriods.jsx [hunk 21]
```diff
@@ -292,38 +427,72 @@ const ReviewPeriods = ({ onPeriodSelected, mode }) => {
...
+                onChange={handleReviewTemplateChange}
+                required
+                value={
+                  periodToAdd.reviewTemplateId
+                    ? periodToAdd.reviewTemplateId
+                    : ''
+                }
+              >
+                <MenuItem key="empty" value="">
+                  -- Please Select --
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/periods/ReviewPeriods.jsx:452]: Is this standard text we want to display in all drop-downs when no item has been selected yet?

File: web-ui/src/components/reviews/periods/ReviewPeriods.jsx [hunk 22]
```diff
@@ -292,38 +427,72 @@ const ReviewPeriods = ({ onPeriodSelected, mode }) => {
...
+            onChange={e =>
+              setPeriodToAdd({...periodToAdd, name: e.target.value})
             }
           />
+          <div style={{margin: '1rem 0'}}>
+            <FormControl fullWidth>
+              <InputLabel id="template-label" required>
+                Review Template
+              </InputLabel>
+              <Select
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/periods/ReviewPeriods.jsx:440]: This is the first new Select component.

File: web-ui/src/components/reviews/periods/ReviewPeriods.jsx [hunk 23]
```diff
@@ -292,38 +427,72 @@ const ReviewPeriods = ({ onPeriodSelected, mode }) => {
...
+                  </MenuItem>
+                ))}
+              </Select>
+            </FormControl>
+          </div>
+          <FormControl fullWidth>
+            <InputLabel id="template-label" required>
+              Self-Review Template
+            </InputLabel>
+            <Select
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/periods/ReviewPeriods.jsx:466]: This is the second new Select component.

**Comment by S78901** [comment]
I see a failing gradle build. May want to check into that.

File: web-ui/.prettierrc [hunk 24]
```diff
@@ -0,0 +1,6 @@
...
+  "arrowParens": "avoid",
+  "bracketSpacing": false,
+  "singleQuote": true,
```
Comment (S78901) [inline web-ui/.prettierrc:4]: Sometimes single vs double quotes are necessary between operating systems is my understanding. Maybe best to not auto-adjust this one?

File: web-ui/.prettierrc [hunk 25]
```diff
@@ -0,0 +1,6 @@
+{
+  "arrowParens": "avoid",
+  "bracketSpacing": false,
```
Comment (S78901) [inline web-ui/.prettierrc:3]: May be just a me thing, I sometimes find value in the extra space for imports to increase readability.

File: web-ui/src/pages/ReviewsPage.jsx (hunk 5 above)
Comment (S78901) [reply web-ui/src/pages/ReviewsPage.jsx:21]: Good call.

File: web-ui/src/components/reviews/periods/ReviewPeriods.jsx (hunk 11 above)
Comment (S78901) [reply web-ui/src/components/reviews/periods/ReviewPeriods.jsx:80]: Is there a reason for the extra line space that is line 79?

File: web-ui/src/components/reviews/periods/ReviewPeriods.jsx (hunk 12 above)
Comment (S78901) [reply web-ui/src/components/reviews/periods/ReviewPeriods.jsx:112]: Nice.

File: web-ui/src/components/reviews/periods/ReviewPeriods.jsx (hunk 21 above)
Comment (S78901) [reply web-ui/src/components/reviews/periods/ReviewPeriods.jsx:452]: Works for me!

File: web-ui/.prettierrc (hunk 24 above)
Comment (mvolkmann) [reply web-ui/.prettierrc:4]: Prettier handles that case. It knows to surround strings that contain single quotes with double and vice versa.

File: web-ui/src/components/reviews/periods/ReviewPeriods.jsx (hunk 11 above)
Comment (mvolkmann) [reply web-ui/src/components/reviews/periods/ReviewPeriods.jsx:80]: I was just highlighting that the three CSS properties I added go together. Removing any one of them loses the effect I'm after.

File: web-ui/src/components/reviews/periods/ReviewPeriods.jsx (hunk 21 above)
Comment (mkimberlin) [reply web-ui/src/components/reviews/periods/ReviewPeriods.jsx:454]: Fine by me as well.

**Pull request merged**